### PR TITLE
run: fix generated unit name clash after soft-reboot

### PR DESCRIPTION
**Pull request by bluca**
When sd-run connects to D-Bus rather than the private socket, it will generate the transient unit name using the bus ID assigned by the D-Bus broker/daemon. The issue is that this ID is only unique per D-Bus run, if the broker/daemon restarts it starts again from 1, and it's a simple incremental counter for each client.
So if a transient unit run-u6.service starts and fails, and it is not collected (default on failure), and the system soft-reboots, any new transient unit might conflict as the counter will restart:

Failed to start transient service unit: Unit run-u6.service was already loaded or has a fragment file.

Get the soft-reboot counter, and if it's greater than zero, append it to the autogenerated unit name to avoid clashes.

### CODE REVIEW

File: src/run/run.c [hunk 1]
```diff
@@ -1313,6 +1313,8 @@ static int transient_timer_set_properties(sd_bus_message *m) {
 }
 
 static int make_unit_name(sd_bus *bus, UnitType t, char **ret) {
+        char soft_reboot_suffix[DECIMAL_STR_MAX(unsigned) + 2] = {0}; /* Add space for '-s' */
```
Comment (YHNdnzj) [inline src/run/run.c]:
```suggestion
        char soft_reboot_suffix[DECIMAL_STR_MAX(unsigned) + 2] = ""; /* Add space for '-s' */
```

File: src/run/run.c [hunk 2]
```diff
@@ -1351,7 +1353,21 @@ static int make_unit_name(sd_bus *bus, UnitType t, char **ret) {
...
+        /* The unique D-Bus names are actually unique per D-Bus instance, so on soft-reboot they will wrap
+         * and start over since the D-Bus broker is restarted. If there's a failed unit left behind that
+         * hasn't been garbage collected, we'll conflict. Append the soft-reboot counter to avoid clashing. */
+        _cleanup_(sd_bus_error_free) sd_bus_error error = SD_BUS_ERROR_NULL;
+        r = bus_get_property_trivial(
+                        bus, bus_systemd_mgr, "SoftRebootsCount", &error, 'u', &soft_reboots_count);
+        if (r < 0)
+                log_debug_errno(r, "Failed to get SoftRebootsCount property, ignoring: %s", bus_error_message(&error, r));
+        else if (soft_reboots_count > 0) {
+                r = snprintf(soft_reboot_suffix, sizeof(soft_reboot_suffix), "-s%u", soft_reboots_count);
```
Comment (YHNdnzj) [inline src/run/run.c]: xsprintf?

Or, maybe just use asprintf below?

**Review by YHNdnzj — COMMENTED** [review]
Hmm, I wonder if we should just always use UUID? Is there a specific reason to try to derive the name from bus first?

**Comment by bluca** [comment]
> Hmm, I wonder if we should just always use UUID? Is there a specific reason to try to derive the name from bus first?

Not sure. It's nice to have an incremental counter in some ways. But I don't want to change behaviour for normal cases, as someone might be relying on it.

**Review by yuwata — APPROVED** [review]

File: src/run/run.c [hunk 3]
```diff
@@ -1351,9 +1352,23 @@ static int make_unit_name(sd_bus *bus, UnitType t, char **ret) {
...
 
-        p = strjoin("run-u", id, ".", unit_type_to_string(t));
-        if (!p)
-                return log_oom();
+        /* The unique D-Bus names are actually unique per D-Bus instance, so on soft-reboot they will wrap
+         * and start over since the D-Bus broker is restarted. If there's a failed unit left behind that
+         * hasn't been garbage collected, we'll conflict. Append the soft-reboot counter to avoid clashing. */
+        _cleanup_(sd_bus_error_free) sd_bus_error error = SD_BUS_ERROR_NULL;
+        r = bus_get_property_trivial(
+                        bus, bus_systemd_mgr, "SoftRebootsCount", &error, 'u', &soft_reboots_count);
```
Comment (poettering) [inline src/run/run.c:1360]: querying this on the --user instance is really weird...

Comment (bluca) [reply src/run/run.c:1360]: https://github.com/systemd/systemd/pull/32127

**Pull request merged**